# Aula 3 - Exercícios - Pandas, Fontes de dados e Agregação
### Consulta a internet (Google, Stackoverflow e etc) liberada!!

In [6]:
# Importe as bibliotecas pandas, numpy, glob e sqlalchemy
import pandas as pd
import numpy as np
import sqlalchemy as sqla
import os


### 3.1 Carregue o arquivo do IDEB 2017 no formato de um DataFrame.  Crie uma função para calcular a média do IDEB de todos os municípios de um determinado Estado. 
#### Remova os registros cujo valor na coluna REDE sejam Municipal, Estadual ou Federal. Deixe somente os registros cujos valores na coluna REDE sejam Pública.


In [70]:
# Dica 1: Antes de calcular, substitua valores '-' por np.nan (NA) com o metodo replace.
# Dica 2: Use a função loc para fazer o filtro de registros.
# Dica 3: Dentre várias maneiras possíveis, uma maneira bastante concisa é usar o operador & (and) para
# congregar dois critérios de filtro (seleção) de registros.
def media_ideb(df, sigla_estado):
    """retorne o valor da media do IDEB contido na coluna col do Estado especificado por sigla_estado.
    """
    # YOUR CODE HERE
    return round(df.loc[(ideb['SG_UF'] == sigla_estado) & (df['REDE'] == 'Pública')]["IDEB12_17"].mean(),2)
    raise NotImplementedError()


In [78]:
path_ideb='../../curso_ciencia_dados2020/data/originais/ideb/ideb_municipios2017.xlsx'
df_ideb = pd.read_excel(path_ideb,skiprows=9, skipfooter=3).replace("-",np.nan)


In [80]:
## Faça seus testes aqui 

# df_ideb = round(ideb.loc[(ideb['SG_UF'] == 'GO') & (ideb['REDE'] == 'Pública')]["IDEB12_17"].mean(),2)
# df_ideb = ideb.groupby((ideb['SG_UF'] == 'DF') & (ideb['REDE'] == 'Pública')).mean()


print (media_ideb(df_ideb, 'SP'))

4.04


In [76]:
# Validação para nota
assert media_ideb(df_ideb, 'DF') == 3.4
assert round(media_ideb(df_ideb, 'SP'),2) == 4.04
assert round(media_ideb(df_ideb, 'GO'),2) == 4.24

### 3.2 Escreva uma função para salvar um DataFrame num banco de dados SQLite  

In [103]:

# Dica: A resolução deste exercício é bem semelhante ao código apresentado na Aula

def save_to_sqlite(df, table_name, con_str):
    """salve o dataframe df num banco de dados sqlite especificado por db_path
    """
    # YOUR CODE HERE
    df.index.names = ['index']
    df.to_sql(table_name,con_str , if_exists = 'replace')
    return
    raise NotImplementedError()


In [104]:
## Faça seus testes aqui
table_name = 'idh' # Escolha o nome da tabela
path_idhm='../../curso_ciencia_dados2020/data/originais/idhm/idhm2013.xlsx'
con_str='sqlite:////home/jovyan/work/curso_ciencia_dados2020/data/processados/idhm/idhm2013.sqlite'
df_idhm = pd.read_excel(path_idhm, sheet_name="IDHM")

# Antes de salvar no banco de dados, defina o nome do índice como index. Conforme exemplo abaixo
#df_idhm.index.names = ['index']
# Validação para nota
def load_table(table, con_str):
    con = sqla.create_engine(con_str)
    return pd.read_sql_query("SELECT * from {0}".format(table), con, index_col='index')
    


df_idhm.index.names = ['index']
save_to_sqlite(df_idhm, table_name, con_str) # Função que deve ser implementada pelo aluno
idhm_from_db = load_table(table_name, con_str)

assert idhm_from_db.equals(df_idhm)


In [98]:
# engine = sqla.create_engine(con_str, echo=True)
# sqlite_connection = engine.connect()
idhm_from_db = None
df_idhm.to_sql(table_name,con_str , if_exists = 'replace')
def load_table(table, con_str):
    con = sqla.create_engine(con_str)
    return pd.read_sql_query("SELECT * from {0}".format(table), con, index_col='index')

idhm_from_db = load_table(table_name, con_str)
idhm_from_db

ANO  UF  Codmun6  Codmun7              Município  ESPVIDA  FECTOT  \
index                                                                       
0      1991  11   110001  1100015  ALTA FLORESTA D'OESTE    62.01    4.08   
1      1991  11   110002  1100023              ARIQUEMES    66.02    3.72   
2      1991  11   110003  1100031                 CABIXI    63.16    3.89   
3      1991  11   110004  1100049                 CACOAL    65.03    3.81   
4      1991  11   110005  1100056             CEREJEIRAS    62.73    3.55   
...     ...  ..      ...      ...                    ...      ...     ...   
16690  2010  52   522200  5222005             VIANÓPOLIS    74.10    2.10   
16691  2010  52   522205  5222054         VICENTINÓPOLIS    73.25    2.29   
16692  2010  52   522220  5222203               VILA BOA    74.07    2.64   
16693  2010  52   522230  5222302          VILA PROPÍCIO    73.09    2.67   
16694  2010  53   530010  5300108               BRASÍLIA    77.35    1.75   

       MORT1  MORT5  RAZDEP  ...  T_SLUZ  pesourb      POP     POPT  \
index                        ...                                      
0      45.58  58.05   73.50  ...   63.20     9379    23417    22835   
1      32.39  41.41   69.97  ...   26.72    45283    56061    55018   
2      41.52  52.94   77.23  ...   74.96     1916     7601     5846   
3      35.37  45.19   71.11  ...   42.66    42689    69173    66534   
4      43.00  54.82   71.09  ...   43.64    14055    19451    19030   
...      ...    ...     ...  ...     ...      ...      ...      ...   
16690  14.40  17.04   45.11  ...    0.17     9170    12501    12445   
16691  15.70  18.51   42.38  ...    0.30     6321     7288     7239   
16692  14.50  17.09   50.80  ...    0.22     3502     4685     4531   
16693  15.90  18.80   50.75  ...    2.41     1504     4739     4523   
16694  14.01  15.85   40.14  ...    0.09  2482210  2543802  2541714   

       I_ESCOLARIDADE  I_FREQ_PROP   IDHM  IDHM_E  IDHM_L  IDHM_R  
index                                                              
0               0.117        0.109  0.329   0.112   0.617   0.516  
1               0.200        0.199  0.432   0.199   0.684   0.593  
2               0.111        0.106  0.309   0.108   0.636   0.430  
3               0.176        0.169  0.407   0.171   0.667   0.593  
4               0.152        0.175  0.386   0.167   0.629   0.547  
...               ...          ...    ...     ...     ...     ...  
16690           0.432        0.747  0.712   0.622   0.818   0.710  
16691           0.389        0.676  0.684   0.562   0.804   0.707  
16692           0.431        0.594  0.647   0.534   0.818   0.619  
16693           0.330        0.624  0.634   0.505   0.802   0.629  
16694           0.723        0.751  0.824   0.742   0.873   0.863  

[16695 rows x 46 columns]

In [100]:
# Validação para nota
def load_table(table, con_str):
    con = sqla.create_engine(con_str)
    return pd.read_sql_query("SELECT * from {0}".format(table), con, index_col='index')
    
path_idhm='../../../curso_ciencia_dados2020/data/originais/idhm/idhm2013.xlsx'
con_str='sqlite:////home/jovyan/work/curso_ciencia_dados2020/data/processados/idhm/idhm2013.sqlite'
table_name='IDHM'
df_idhm = pd.read_excel(path_idhm, sheet_name="IDHM")

df_idhm.index.names = ['index']
save_to_sqlite(df_idhm, table_name, con_str) # Função que deve ser implementada pelo aluno
idhm_from_db = load_table(table_name, con_str)

assert idhm_from_db.equals(df_idhm)

FileNotFoundError: [Errno 2] No such file or directory: '../../../curso_ciencia_dados2020/data/originais/idhm/idhm2013.xlsx'

### 3.3 Escreva uma função executar uma query SQL de agregação num banco de dados SQLite que retorne a média dos valores de uma coluna VALORES agregados por uma coluna CATEGORIAS.
#### Retorne a coluna CATEGORIA na tabela de resultados para identificar os valores. Ordene os resultados pelo valor da média em ordem decrescente.
#### Exemplo: Calcule média do IDHM (coluna VALORES) agregado por Ano da pesquisa (CATEGORIAS).

In [ ]:
# Dica: A resposta deste exercício é bem semelhante ao código mostrado na aula
def soma_com_agregacao(val, cat, table, con_str):
    """retorne um dataframe contendo a soma dos valores em val dentro das categorias cat da tabela table
    """
    # YOUR CODE HERE
    raise NotImplementedError()


In [ ]:
## Faça seus testes aqui



In [ ]:
# Validação para nota
con_str='sqlite:////home/jovyan/work/curso_ciencia_dados2020/data/processados/idhm/idhm2013.sqlite'
idhm_anos = pd.DataFrame({'ANO': {0: 2010, 1: 2000, 2: 1991}, 'Media': {0: 0.6591574123989226, 1: 0.5234808625336926, 2: 0.38137232704402557}})
assert soma_com_agregacao('IDHM', 'ANO', 'IDHM', con_str).equals(idhm_anos)

### 3.4 Escreva uma função para fazer o merge dos dataframes de população e de PIB de 2017. A função deve receber como parâmetro o caminho para cada um dos dataframes.

In [ ]:
# Dica 1: A coluna que contem o código IBGE do município deve ter o mesmo formato (string) e a mesma
#         quantidade de caracteres nas duas tabelas.
# Dica 2: Converta o codigo do município para string ao carregar o dataframe de população utilizando
#         o mapeamento de tipos dtype={'cod_munic': str, 'cod_uf': str}. Assim o codigo do municipio não perde
#         os zeros a esquerda.
# Dica 3: Concatene o codigo da UF com o código do município usando o operador + e salve em outra coluna.
# Dica 4: Use o código apresentado na aula para carregar o dataframe do PIB de 2017, especifique o dtype da
#         coluna que contém o codigo do municipio.
# Dica 5: Renomeie as colunas apropriadamente.
# Dica 6: Depois que os dois dataframes tiverem o codigo ibge (com valores equivalentes
#         e mesmo tipo de dados) você pode fazer o merge.

def merge_pop_pib(path_pop, path_pib):
    """retorne um dataframe resultante do merge dos dados do pib e da população
    """
    # YOUR CODE HERE
    raise NotImplementedError()


In [ ]:
## Faça seus testes aqui
path_pib = '../../../curso_ciencia_dados2020/data/originais/pib/pib_municipios.xlsx'
path_pop='../../../curso_ciencia_dados2020/data/originais/populacao/estimativa_dou_2017.xlsx'
df_pib_pop = merge_pop_pib(path_pop, path_pib)
df_pib_pop.head()

In [ ]:
# Validação para nota
test_merge = {'pop2017': 25437, '2017': 498864}
df_pib_pop = merge_pop_pib(path_pop, path_pib)
assert df_pib_pop.loc[0,['pop2017', '2017']].to_dict() == test_merge 

### 3.5 Escreva uma função para calcular o PIB percapita de cada município do Brasil, considerando o dataframe do exercício anterior.

In [ ]:
# Dica 1
def pib_percapita(pib_pop):
    """retorne retorne um dataframe contendo o codigo ibge de 7 caracteres,
       o nome do municipio e seu PIB percapita.
    """
    # YOUR CODE HERE
    raise NotImplementedError()


In [ ]:
## Faça seus testes aqui



In [ ]:
# Validação para nota
pib_percapita_2017 = pib_percapita(df_pib_pop)
assert round(pib_percapita(df_pib_pop).loc[0,'pib_percapita'],2) == 19.61

### 3.6 Escreva uma função para retornar os registros dos 10 maiores PIBs percapita do brasil a partir do dataframe do exercício anterior. 
#### Escreva uma função genérica que receba como parâmetro um dataframe e o nome da coluna de ordenação


In [ ]:
# Dica 1: Use o método sort_values para ordenar o dataframe. Para uma ordenação descendente use ascending=False.
# Dica 2: Use o atributo iloc para retornar todas as colunas dos 10 primeiros registros.
def top10(df, col):
    """retorne os top 10 registros do dataframe df ordenados por col
    """
    # YOUR CODE HERE
    raise NotImplementedError()


In [ ]:
## Faça seus testes aqui


In [ ]:
# Validação para nota
test_data = {'cod_ibge7': {3677: '3536505', 3547: '3524709'},
             'municipio': {3677: 'Paulínia', 3547: 'Jaguariúna'},
             'uf': {3677: 'SP', 3547: 'SP'},
             'pib_percapita': {3677: 344.8471692406755, 3547: 209.32086192900894}}

assert top10(pib_percapita_2017, 'pib_percapita').iloc[[0,9]].to_dict() == test_data 

### 3.7 Escreva uma função para contar quantos nomes de municípios são diferentes nas bases de dados do PIB e da população

In [ ]:
def compare_df_cols(df, col1, col2):
    """retorne a quantidade de registros em que o valor do registro na coluna col1 difere do valor em col2
    """
    # YOUR CODE HERE
    raise NotImplementedError()


In [ ]:
## Faça seus testes aqui


In [ ]:
# Validação para nota
assert compare_df_cols(df_pib_pop, 'municipio', 'municipio_y') == 0

### 3.8 Considere algumas linhas de uma base de dados de morbidade apresentadas a seguir. Perceba que há dados de vários meses de 2020 para cada Estado da Federação.
#### Escreva uma função para pivotar esta base de dados apresentando os meses do ano nas colunas, os municípios nas linhas e a quantidade de óbitos como valores das células. 

|       | Município               |   cod_ibge6 | mes   |   ano | mes_ano    | Óbitos   |
|------:|:------------------------|------------:|:------|------:|:-----------|:---------|
|     0 | Alta Floresta D'Oeste   |      110001 | Abr   |  2020 | 2020-04-01 | 8        |
|     1 | Alto Alegre dos Parecis |      110037 | Abr   |  2020 | 2020-04-01 | 1        |
|     2 | Alto Paraíso            |      110040 | Abr   |  2020 | 2020-04-01 | 2        |
| 33347 | Vila Boa                |      522220 | Jan   |  2020 | 2020-01-01 | -        |
| 33348 | Vila Propício           |      522230 | Jan   |  2020 | 2020-01-01 | -        |
| 33349 | Brasília                |      530010 | Jan   |  2020 | 2020-01-01 | 578      |

In [ ]:
# Dica 1: Se aparecer um erro (uma exceção) com este texto "ValueError: Index contains duplicate entries, cannot reshape"
#       é porque o seu índice contém valores duplicados. 
# Dica 2: Se você usar o atributo 'mes_ano' o pandas já ordena de forma cronológica as colunas.
def pivot_morbidade(file_path):
    """retorne um dataframe pivotado
    """
    # YOUR CODE HERE
    raise NotImplementedError()


In [ ]:
## Faça seus testes aqui
file_path = '../../../curso_ciencia_dados2020/data/originais/morbidade/consolidado/morbidade_simplificado.csv'


In [ ]:
# Validação para nota
file_path = '../../../curso_ciencia_dados2020/data/originais/morbidade/consolidado/morbidade_simplificado.csv'
test_data = {'2020-01-01': '5', '2020-02-01': '2', '2020-03-01': '3', '2020-04-01': '8', '2020-05-01': '2', 
             '2020-06-01': '2'}
assert pivot_morbidade(file_path).iloc[0].to_dict() == test_data
